### PR TITLE
New bounds

### PR DESCRIPTION
**Pull request by janosg**
This PR implements the changes for specifying bound constraints on parameters from the typing proposal and aligns the bounds argument with scipy.

### CODE REVIEW

File: src/optimagic/optimization/optimize.py [hunk 1]
```diff
@@ -1,3 +1,16 @@
+"""Public functions for optimization.
+
+This module defines the public functions `maximize` and `minimize` that will be called
+by users.
+
+Internally, `maximize` and `minimize` just pass all arguments through to `_optimize`
+and add a `direction` argument.
+
+`optimize` processes the input arguments and converts them to stricter types. It then
+calls `_internal_optimize` which does the actual optimization.
```
Comment (janosg) [inline src/optimagic/optimization/optimize.py:10]: This is already outdated.

File: src/optimagic/parameters/bounds.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    specified (in the case where params is a DataFrame with bounds as a column).
+
+    Args:
+        params: The parameter pytree.
+        bounds: The lower and upper bounds.
+        registry: pybaum registry.
+
+    Returns:
+        Consolidated and flattened lower_bounds.
+        Consolidated and flattened upper_bounds.
```
Comment (janosg) [inline src/optimagic/parameters/bounds.py:98]: @timmens check the docstring

File: src/optimagic/deprecations.py [hunk 3]
```diff
@@ -95,3 +96,33 @@ def replace_and_warn_about_deprecated_algo_options(algo_options):
...
+    upper_bounds,
+    bounds,
+    soft_lower_bounds=None,
+    soft_upper_bounds=None,
+):
+    old_bounds = {
+        "lower": lower_bounds,
+        "upper": upper_bounds,
+        "soft_lower": soft_lower_bounds,
+        "soft_upper": soft_upper_bounds,
```
Comment (timmens) [inline src/optimagic/deprecations.py:112]: Should be `"lower_bounds"`, `"upper_bounds"`, `"soft_lower_bounds"`, ..., right?

File: src/optimagic/deprecations.py [hunk 4]
```diff
@@ -95,3 +96,33 @@ def replace_and_warn_about_deprecated_algo_options(algo_options):
...
+    if old_present:
+        msg = (
+            f"Specifying bounds via the arguments {', '.join(old_present)} is "
+            "deprecated and will be removed in optimagic version 0.6.0 and later. "
+            "Please use the `bounds` argument instead."
+        )
+        warnings.warn(msg, FutureWarning)
+
+    if bounds is None and old_present:
+        bounds = Bounds(**old_bounds)
```
Comment (timmens) [inline src/optimagic/deprecations.py:126]: What about the case where both the new and old bounds are set. In the current implementation we use the new bounds, but do not throw a warning other than the FutureWarning above.

File: src/optimagic/parameters/bounds.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+from optimagic.parameters.tree_registry import get_registry
+from dataclasses import dataclass
+from optimagic.typing import PyTree, PyTreeRegistry
+from scipy.optimize import Bounds as ScipyBounds
+from typing import Sequence
+from numpy.typing import NDArray
+from typing import Any
+
+
+@dataclass
```
Comment (timmens) [inline src/optimagic/parameters/bounds.py:17]: Not frozen with replace method?

**Review by timmens — CHANGES_REQUESTED** [review]
Very nice PR. Only a few minor questions.

File: src/optimagic/deprecations.py (hunk 3 above)
Comment (janosg) [reply src/optimagic/deprecations.py:112]: No, but I need to adjust for it in the error message.

File: src/optimagic/deprecations.py (hunk 4 above)
Comment (janosg) [reply src/optimagic/deprecations.py:126]: As discussed, this behavior is safe enough.

Comment (janosg) [reply src/optimagic/deprecations.py:126]: Same.

File: src/optimagic/parameters/bounds.py (hunk 5 above)
Comment (janosg) [reply src/optimagic/parameters/bounds.py:17]: Should be frozen, will change it.

**Review by timmens — APPROVED** [review]

**Pull request merged**